src/Security/Authentication/Negotiate/src/Internal/LdapAdapter.cs

```diff
@@ -116,10 +140,51 @@ private static void GetNestedGroups(LdapConnection connection, ClaimsIdentity pr
             var group = searchResponse.Entries[0]; // Get the object that was found on ldap
             var groupDN = group.DistinguishedName;
 
+            if (processedGroups.Contains(groupDN)) {
+                //No need to continue, this group was resolved before
+                return;
+            }
+
+            retrievedClaims.Add(new KeyValuePair<string, string>(principal.RoleClaimType, groupCN));
             processedGroups.Add(groupDN);
 
+            //Get the group SID
+            var groupSID = group.Attributes["objectsid"];
+            if (groupSID is not null)
+            {
+                if (groupSID.Count == 1)
+                {
+                    //For some reason it is sometimes string and sometimes byte[]
+                    //when it is returned as a string then every byte is converted to a char and simply put together as a string
+                    switch (groupSID[0])
+                    {
+                        case string groupSIDstr:
+                            ReadOnlySpan<char> groupSIDspn = groupSIDstr;
+                            Span<byte> lgroupSIDba = stackalloc byte[groupSIDspn.Length];
+                            for (int i = 0; i < groupSIDspn.Length; ++i)
+                            {
+                                byte[] bytes = BitConverter.GetBytes(groupSIDspn[i]);
+                                lgroupSIDba[i] = bytes[0];
```

Comment:
Can you use the span overload (`TryWriteBytes`) to avoid the allocation of that temp array?

Reply:
The problem is that this should not be a string at all but rather a byte array. I don't know why this happens for builtin groups, but it turns the byte array of the SID `0x01 0x02 0x03 0x04 ...` into a completely nonsensical string (char[]) `0x0001 0x0002 0x0003 0x0004`. I am not sure if this happens endian safe or as a raw copy. I'll try to change this to avoid the temporary array and make the stackalloc safer by using a static size and slice method.